pkg/services/meta/notifications.go

```diff
@@ -0,0 +1,571 @@
+	}
+
+	commPrefix := e.oID[:]
+
+	st.opsBatch[string(append([]byte{oidIndex}, commPrefix...))] = []byte{}
+	st.opsBatch[string(append([]byte{sizeIndex}, commPrefix...))] = e.size.Bytes()
+	if len(e.firstObject) == sha256.Size {
+		st.opsBatch[string(append([]byte{firstPart}, commPrefix...))] = e.firstObject
+	}
+	if len(e.prevObject) == sha256.Size {
+		st.opsBatch[string(append([]byte{previousPart}, commPrefix...))] = e.prevObject
+	}
+	if len(e.deletedObjects) > 0 {
+		st.opsBatch[string(append([]byte{deletedIndex}, commPrefix...))] = e.deletedObjects
+	}
+	if len(e.lockedObjects) > 0 {
+		st.opsBatch[string(append([]byte{lockedIndex}, commPrefix...))] = e.lockedObjects
+	}
+	if e.typ != objectsdk.TypeRegular {
+		st.opsBatch[string(append([]byte{typeIndex}, commPrefix...))] = []byte{byte(e.typ)}
+	}
+
+	// TODO: receive full object header and store its non-MTP parts in
+	//  persistent object storage
+
+	return nil
+}
+
+type cnrEvent struct {
+	cID cid.ID
+}
+
+func parseCnrNotification(ev *state.ContainedNotificationEvent) (cnrEvent, error) {
+	var res cnrEvent
+
+	arr, ok := ev.Item.Value().([]stackitem.Item)
+	if !ok {
+		return res, fmt.Errorf("unexpected notification stack item: %T", ev.Item.Value())
+	}
+
+	switch ev.Name {
+	case cnrDeleteName:
+		const expectedNotificationArgs = 1
+		if len(arr) != expectedNotificationArgs {
+			return res, fmt.Errorf("unexpected number of items on stack: %d, expected: %d", len(arr), expectedNotificationArgs)
+		}
+	case cnrPutName:
+		const expectedNotificationArgs = 2
+		if len(arr) != expectedNotificationArgs {
+			return res, fmt.Errorf("unexpected number of items on stack: %d, expected: %d", len(arr), expectedNotificationArgs)
+		}
+	}
+
+	cID, ok := arr[0].Value().([]byte)
+	if !ok {
+		return res, fmt.Errorf("unexpected container ID stack item: %T", arr[0].Value())
+	}
+	if len(cID) != sha256.Size {
+		return res, fmt.Errorf("unexpected container ID len: %d", len(cID))
+	}
+
+	return cnrEvent{cID: cid.ID(cID)}, nil
+}
+
+func (m *Meta) dropContainer(cID cid.ID) error {
+	m.m.Lock()
+	defer m.m.Unlock()
+
+	st, ok := m.storages[cID]
+	if !ok {
+		return nil
+	}
+
+	err := st.drop()
+	if err != nil {
+		m.l.Warn("drop container %s: %w", zap.Stringer("cID", cID), zap.Error(err))
+	}
+
+	delete(m.storages, cID)
+
+	return nil
+}
+
+func parseEpochNotification(ev *state.ContainedNotificationEvent) (int64, error) {
+	const expectedNotificationArgs = 1
+
+	arr, ok := ev.Item.Value().([]stackitem.Item)
+	if !ok {
+		return 0, fmt.Errorf("unexpected notification stack item: %T", ev.Item.Value())
+	}
+	if len(arr) != expectedNotificationArgs {
+		return 0, fmt.Errorf("unexpected number of items on stack: %d, expected: %d", len(arr), expectedNotificationArgs)
+	}
+
+	epoch, ok := arr[0].Value().(*big.Int)
+	if !ok {
+		return 0, fmt.Errorf("unexpected epoch stack item: %T", arr[0].Value())
+	}
+
+	return epoch.Int64(), nil
+}
+
+func (m *Meta) handleEpochNotification(e int64) error {
```

Comment:
Expired objects are not deleted.

Reply:
hm, we do not have expiration epoch in meta-from-chain, so we cannot find such cases without object HEADing object. or we have to add it to the contract

Reply:
Can be postponed.